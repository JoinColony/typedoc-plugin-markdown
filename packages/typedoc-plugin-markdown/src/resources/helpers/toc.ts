import * as Handlebars from 'handlebars/runtime';
import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionGroup,
} from 'typedoc';
import { HelperOptions } from '../../utils/models';

export const toc = (
  props: ProjectReflection | DeclarationReflection,
  options: HelperOptions,
) => {
  const md: string[] = [];

  const hideInPageTOC = options.data?.options.hideInPageTOC;

  const isVisible = props.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );

  function pushGroup(group: ReflectionGroup, md: string[]) {
    const children = group.children.map(
      (child) =>
        `- [${Handlebars.helpers.escapeChars(
          child.name,
        )}](${Handlebars.helpers.relativeURL(child.url)})`,
    );
    md.push(children.join('\n'));
  }

  if ((!hideInPageTOC && props.groups) || (isVisible && props.groups)) {
    if (!hideInPageTOC) {
      md.push(`## Table of contents\n\n`);
    }
    const headingLevel = hideInPageTOC ? `##` : `###`;
    props.groups?.forEach((group) => {
      const groupTitle = group.title;
      if (group.categories) {
        group.categories.forEach((category) => {
          md.push(`${headingLevel} ${category.title} ${groupTitle}\n\n`);
          pushGroup(category as any, md);
          md.push('\n');
        });
      } else {
        if (!hideInPageTOC || group.allChildrenHaveOwnDocument()) {
          md.push(`${headingLevel} ${groupTitle}\n\n`);
          pushGroup(group, md);
          md.push('\n');
        }
      }
    });
  }
  return md.length > 0 ? md.join('\n') : null;
};
