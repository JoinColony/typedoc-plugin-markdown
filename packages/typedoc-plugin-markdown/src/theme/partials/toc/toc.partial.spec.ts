import { DeclarationReflection, ProjectReflection, Reflection } from 'typedoc';
import { tocPartial } from './toc.partial';

describe(`TOC Partial:`, () => {
  let project: ProjectReflection;
  let moduleReflection: Reflection;
  let classReflection: Reflection;

  beforeAll(async () => {
    project = global.getProject('toc.ts');
    moduleReflection = project;
    classReflection = project.findReflectionByName('SomeClass');
  });

  describe(`(default)`, () => {
    test(`should display toc for module'`, () => {
      expect(
        tocPartial(
          global.getMockContext(),
          moduleReflection as DeclarationReflection,
        ),
      ).toMatchSnapshot();
    });

    test(`should display toc for class'`, () => {
      expect(
        tocPartial(
          global.getMockContext(),
          classReflection as DeclarationReflection,
        ),
      ).toMatchSnapshot();
    });
  });
  describe(`(hideInPageToc)`, () => {
    test(`should not display toc for class'`, () => {
      expect(
        tocPartial(
          global.getMockContext({ hideInPageTOC: true }),
          classReflection as DeclarationReflection,
        ),
      ).toEqual('');
    });
  });
});
