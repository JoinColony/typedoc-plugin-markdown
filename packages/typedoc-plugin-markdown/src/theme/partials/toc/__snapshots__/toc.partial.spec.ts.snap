// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOC Partial: (default) should display toc for class' 1`] = `
"## Table of contents


### Constructors


- [constructor](( relativeURL ))


### Methods


- [someMethod](( relativeURL ))

"
`;

exports[`TOC Partial: (default) should display toc for module' 1`] = `
"## Table of contents


### Classes


- [SomeClass](( relativeURL ))


### Type aliases


- [answer](( relativeURL ))


### Variables


- [answer](( relativeURL ))

"
`;
